crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,529 @@
+                        }
+                    }
+                }
+
+                Update::RemoveChunk(expected_chunk_identifier) => {
+                    let chunk_index = self
+                        .chunks
+                        .iter()
+                        .position(|(chunk_identifier, _)| {
+                            chunk_identifier == expected_chunk_identifier
+                        })
+                        // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and assuming
+                        // `Self::chunks` is correctly initialized, it is not possible to remove a
+                        // chunk that does not exist. If this predicate fails, it means
+                        // `LinkedChunk` or `Updates` contain a bug.
+                        .expect("Removing a chunk: The chunk is not found");
+
+                    // It's OK to ignore the result. The `chunk_index` exists because it's been
+                    // found, and we don't care about its associated value.
+                    let _ = self.chunks.remove(chunk_index);
+                }
+
+                Update::PushItems { position_hint: position, items } => {
+                    let expected_chunk_identifier = position.chunk_identifier();
+
+                    let (chunk_index, offset, chunk_length) = {
+                        let control_flow = self.chunks.iter_mut().enumerate().try_fold(
+                            position.index(),
+                            |offset, (chunk_index, (chunk_identifier, chunk_length))| {
+                                if chunk_identifier == &expected_chunk_identifier {
+                                    ControlFlow::Break((chunk_index, offset, chunk_length))
+                                } else {
+                                    ControlFlow::Continue(offset + *chunk_length)
+                                }
+                            },
+                        );
+
+                        match control_flow {
+                            ControlFlow::Break(value) => Some(value),
+                            ControlFlow::Continue(..) => None,
+                        }
+                    }
+                    // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and assuming
+                    // `Self::chunks` is correctly initialized, it is not possible to push items on
+                    // a chunk that does not exist. If this predicate fails, it means `LinkedChunk`
+                    // or `Updates` contain a bug.
+                    .expect("Pushing items: The chunk is not found");
+
+                    *chunk_length += items.len();
+
+                    // See `mute_push_items` to learn more.
+                    if mute_push_items {
+                        continue;
+                    }
+
+                    // Optimisation: we can emit a `VectorDiff::Append` in this particular case.
+                    if chunk_index == self.chunks.len() - 1 {
```

Comment:
Just to make sure we don't underflow:

```suggestion
                    if chunk_index + 1 == self.chunks.len() {
```

Reply:
Good eye, done!